dynawo/sources/Solvers/Common/DYNSolverImpl.cpp

```diff
@@ -245,73 +245,54 @@ Solver::Impl::solve(double tAim, double &tNxt) {
 bool
 Solver::Impl::evalZMode(vector<state_g> &G0, vector<state_g> &G1, const double & time) {
   Timer timer("SolverIMPL::evalZMode");
-  bool zChange = false;
-  bool modeChange = false;
-  bool stableRoot = true;
   bool change = false;
 
-  for (int i = 0; i < maxNumberUnstableRoots; ++i) {
-    // evalZ
+  // evalZ part
+  bool nonSilentZChange;
+  int i = 0;
+  do {
+    nonSilentZChange = false;
     model_->evalZ(time);
-    zChange = model_->zChange();
     ++stats_.nze_;
 
-    // evaluate G and compare with previous values
-    stableRoot = detectUnstableRoot(G0, G1, time);
-
-    if (model_->getSilentZChange())
+    if (model_->getSilentZChange()) {
       state_.setFlags(SilentZChange);
-    if (zChange) {
-      change = true;
+    } else if (model_->zChange()) {
+      model_->evalG(time, G1);
+      ++stats_.nge_;
       state_.setFlags(ZChange);
-    } else if (stableRoot) {
-      break;
+      nonSilentZChange = true;
+      change = true;
+#ifdef _DEBUG_
+      std::copy(G1.begin(), G1.end(), G0.begin());
+      printUnstableRoot(G1, G0);
+#endif
     }
-  }
 
-  // evalMode
+    ++i;
+    if (i == maxNumberUnstableRoots)
```

Comment:
better to have it (i>= ...) in case someone modifies the way i is incremented

Reply:
Solved.